### DISCUSSION TITLE
How to implement effects such as pitch changing/vibrato?

### DISCUSSION THREAD

**Discussion by henrikvilhelmberglund** [danigb/smplr#85] [Q&A]
I'm making an lsdj inspired tracker (super rough current state can be found here https://nejime.henrikvilhelmberglund.com/) and I want to implement a CMD column that can have commands for pitch changing and vibrato among others.
Is there any way to do this? I think I would need a LFO oscillator that connects to a DelayNode on a smplr instrument but a bit unsure about how this all works. I looked at the code of the reverb effect but didn't really understand a lot. Maybe this is out of scope for smplr?
Would appreaciate any input! 😊
edit: saw a bunch of effects here: https://github.com/cwilso/Audio-Input-Effects

**Comment by danigb** [comment]
I'd say vibrato would need to be implemented natively in smplr because it needs to connect to audio buffer source directly. However other effects (like https://github.com/Theodeus/tuna) can be added to any instrument in the same we use the reverb. 

A possible (not tested) example:

```ts
const chorus = new tuna.Chorus({
    rate: 1.5,
    feedback: 0.2,
    delay: 0.0045,
    bypass: false
});
const piano = new SplendidGrandPiano(context, { volume });
piano.output.addEffect("chorus", chorus, 0.5);
```

Let me know if you find a way to do it

EDIT: BTW, love to see where you take your tracker to 👍
